client/src/txnmgr.rs

```diff
@@ -0,0 +1,326 @@
+    ///     confirmations,
+    ///     txid,
+    /// }) = txn_mgr.wait(txn_id, std::time::Duration::from_secs(10))
+    /// {}
+    ///```
+    pub fn submit_transaction(
+        &mut self,
+        txn: Transaction,
+        timeout: time::Duration,
+    ) -> Result<Event, handle::Error> {
+        let tx_id = txn.txid();
+
+        let (transmitter, recvr) = chan::bounded(1);
+        self.handle
+            .command(Command::BroadcastTransactionInv(tx_id, transmitter))?;
+
+        let peers = recvr.recv()?;
+        if peers.is_empty() {
+            return Err(Error::RelayPeer.into());
+        }
+
+        let mut txn_map = match self.handle.transactions.write() {
+            Ok(e) => e,
+            Err(e) => return Err(e.into()),
+        };
+
+        let events = self.handle.events();
+        let mut peers_responded = HashSet::new();
+
+        // Keep on listening for peers that want our transaction till timeout.
+        let result = event::wait(
+            &events,
+            |e| match e {
+                event::Event::Received(e, NetworkMessage::GetData(inv)) => {
+                    if inv.contains(&Inventory::Transaction(tx_id)) {
+                        let result = self
+                            .handle
+                            .command(Command::SubmitTransaction(e, txn.clone()));
+                        peers_responded.insert(e);
+                        return Some(result);
+                    }
+
+                    None
+                }
+
+                event::Event::SyncManager(syncmgr::Event::BlockReceived(_, block, height)) => {
+                    self.on_blocks_received(tx_id, block, height, &mut *txn_map)
+                }
+
+                _ => None,
+            },
+            timeout,
+        );
+
+        let peers_responded_len = peers_responded.len();
+
+        // We have broadcasted the tx ID but couldn't send
+        // tx to peers probably due to a timeout.
+        if peers_responded_len == 0 {
+            return Ok(Event::Pending {
+                announcements: peers.len(),
+                txid: tx_id,
+            });
+        }
+
+        match result? {
+            Ok(_) => Ok(Event::Accepted {
+                confirmations: peers_responded_len,
+                txid: tx_id,
+            }),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Wait for a change of event of specified transaction ID.
+    pub fn wait(&self, tx_id: Txid, timeout: time::Duration) -> Result<Event, handle::Error> {
+        let mut txn_map = match self.handle.transactions.write() {
+            Ok(e) => e,
+            Err(e) => return Err(e.into()),
+        };
+
+        let events = self.handle.events();
+
+        let keys: Vec<Txid> = txn_map.keys().cloned().into_iter().collect();
+
+        let result = event::wait(
+            &events,
+            |e| match e {
+                event::Event::Received(e, NetworkMessage::Inv(inv)) => {
```

Comment:
?

Reply:
This should be removed and integrated with the `wait` method.